File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> combination;
        helper(candidates, 0, 0, target, combination, ans);

        return ans;
    }

    void helper(vector<int> nums, int currIdx, int currSum,int target, vector<int> combination, vector<vector<int>>& ans ){

        if (currSum == target){
            ans.push_back(combination);
            return;
        } 
        if (currIdx == nums.size() || currSum > target) return;
        

        // choice 1 choosing same number
        combination.push_back(nums[currIdx]);
        helper(nums, currIdx, currSum + nums[currIdx], target, combination, ans);

        // deleting the element at backtracking
        combination.pop_back();
        // choice 2 chosing next from array
        helper(nums, ++currIdx, currSum, target, combination, ans);

    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();

        f(ans, candidates, target, candidates.length - 1, new ArrayList<Integer>());

        return ans;
    }

    void f(List<List<Integer>> collection, int[] candidates, int target, int idx, List<Integer> subset) {
        if (target == 0) {
            List<Integer> validSubset = new ArrayList<>(subset);
            collection.add(validSubset);
            return;
        }

        if (idx < 0) return;

        // take
        if (candidates[idx] <= target) {
            subset.add(candidates[idx]);    
            f(collection, candidates, target - candidates[idx], idx , subset);
            subset.remove(subset.size() - 1); // backtracking 
        }


        // no take
        f(collection, candidates, target, idx - 1, subset);
    }
}